Reset Password Crash: Open URL in MX Player

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/d6a58a56-683a-44fb-b764-421e1c3e7d0f

```
java.lang.SecurityException: Permission Denial: starting Intent { act=android.intent.action.VIEW dat=http://ankiweb.net/... flg=0x10000 cmp=com.mxtech.videoplayer.ad/com.mxtech.videoplayer.ActivityWebBrowser } from ProcessRecord{d14351dd0 30129:com.ichi2.anki/u0a214} (pid=30129, uid=10214) not exported from uid 10256
at android.os.Parcel.createException(Parcel.java:1966)
at android.os.Parcel.readException(Parcel.java:1934)
at android.os.Parcel.readException(Parcel.java:1884)
at android.app.IActivityManager$Stub$Proxy.startActivity(IActivityManager.java:3604)
at android.app.Instrumentation.execStartActivity(Instrumentation.java:1669)
at android.app.Activity.startActivityForResult(Activity.java:4688)
at androidx.fragment.app.FragmentActivity.startActivityForResult(FragmentActivity.java:6)
at android.app.Activity.startActivityForResult(Activity.java:4646)
at androidx.fragment.app.FragmentActivity.startActivityForResult(FragmentActivity.java:3)
at com.ichi2.anki.AnkiActivity.startActivityForResult(AnkiActivity.java:1)
at android.app.Activity.startActivity(Activity.java:5007)
at android.app.Activity.startActivity(Activity.java:4975)
at com.ichi2.anki.AnkiActivity.startActivityWithoutAnimation(AnkiActivity.java:2)
at com.ichi2.anki.MyAccount.resetPassword(MyAccount.java:4)
at com.ichi2.anki.MyAccount.e(MyAccount.java:1)
at com.ichi2.anki.d4.onClick(Unknown Source:2)
at android.view.View.performClick(View.java:7333)
at android.widget.TextView.performClick(TextView.java:14160)
at android.view.View.performClickInternal(View.java:7299)
at android.view.View.access$3200(View.java:846)
at android.view.View$PerformClick.run(View.java:27772)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:6981)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445)
Caused by: android.os.RemoteException: Remote stack trace:
at com.android.server.am.ActivityStackSupervisor.checkStartAnyActivityPermission(ActivityStackSupervisor.java:2128)
at com.android.server.am.ActivityStarter.startActivity(ActivityStarter.java:889)
at com.android.server.am.ActivityStarter.startActivity(ActivityStarter.java:659)
at com.android.server.am.ActivityStarter.startActivityMayWait(ActivityStarter.java:1503)
at com.android.server.am.ActivityStarter.execute(ActivityStarter.java:601)
#################################################################
```

https://stackoverflow.com/questions/57223127/securityexception-crash-opening-link-in-external-browser-when-mx-player-is-inst

--- a/AnkiDroid/src/main/java/com/ichi2/anki/MyAccount.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/MyAccount.java
@@ -165,13 +165,7 @@ public class MyAccount extends AnkiActivity {
 
 
     private void resetPassword() {
-        if (AdaptionUtil.hasWebBrowser(this)) {
-            Intent intent = new Intent(Intent.ACTION_VIEW);
-            intent.setData(Uri.parse(getResources().getString(R.string.resetpw_url)));
-            startActivityWithoutAnimation(intent);
-        } else {
-            UIUtils.showThemedToast(this, getResources().getString(R.string.no_browser_notification) + getResources().getString(R.string.resetpw_url), false);
-        }
+        super.openUrl(Uri.parse(getResources().getString(R.string.resetpw_url)));
     }
 
 

--- a/AnkiDroid/src/main/java/com/ichi2/compat/customtabs/CustomTabsFallback.java
+++ b/AnkiDroid/src/main/java/com/ichi2/compat/customtabs/CustomTabsFallback.java
@@ -17,13 +17,26 @@ import android.app.Activity;
 import android.content.Intent;
 import android.net.Uri;
 
+import com.ichi2.anki.AnkiDroidApp;
+import com.ichi2.anki.R;
+import com.ichi2.anki.UIUtils;
+
 /**
  * A Fallback that opens a Webview when Custom Tabs is not available
  */
 public class CustomTabsFallback implements CustomTabActivityHelper.CustomTabFallback {
     @Override
     public void openUri(Activity activity, Uri uri) {
-        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
-        activity.startActivity(intent);
+        try {
+            Intent intent = new Intent(Intent.ACTION_VIEW, uri);
+            activity.startActivity(intent);
+        } catch (Exception e) {
+            // This can occur if the provider is not exported: #7721
+            // this should not happen as we don't reach here if there's no valid browser.
+            // and I assume an exported intent will take priority over a non-exported intent.
+            // Add an exception report to see if I'm wrong
+            AnkiDroidApp.sendExceptionReport(e, "CustomTabsFallback::openUri");
+            UIUtils.showThemedToast(activity, activity.getString(R.string.web_page_error, uri), false);
+        }
     }
 }

--- a/AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.java
@@ -86,6 +86,10 @@ public class AdaptionUtil {
         PackageManager pm = context.getPackageManager();
         List<ResolveInfo> list = pm.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
         for (ResolveInfo ri : list) {
+            if (!isValidBrowser(ri)) {
+                continue;
+            }
+
             // If we aren't a restricted device, any browser will do
             if (!isRestrictedLearningDevice()) {
                 return true;
@@ -97,6 +101,12 @@ public class AdaptionUtil {
         }
         // Either there are no web browsers, or we're a restricted learning device and there's no system browsers.
         return false;
+    }
+
+
+    private static boolean isValidBrowser(ResolveInfo ri) {
+        // https://stackoverflow.com/a/57223246/
+        return ri != null && ri.activityInfo != null && ri.activityInfo.exported;
     }
 
 

--- a/AnkiDroid/src/main/res/values/02-strings.xml
+++ b/AnkiDroid/src/main/res/values/02-strings.xml
@@ -257,6 +257,7 @@
     <string name="activity_start_failed">The system does not have an app installed that can perform this action.</string>
     <string name="multimedia_editor_image_compression_failed"><![CDATA[Camera images may be large. You may wish to compress & resize images in the media directory]]></string>
     <string name="no_browser_notification">No browser found，please visit web page by pc or mobile: </string>
+    <string name="web_page_error">Error loading page: %s</string>
     <string name="no_outgoing_link_in_cardbrowser">External page link in card is not supported.</string>
     <!-- The name of the deck which corrupt cards will be moved to -->
     <string name="check_integrity_recovered_deck_name">Recovered Cards</string>

